tidy3d/web/api/container.py

```diff
@@ -176,25 +170,19 @@ def run(self, path: str = DEFAULT_DATA_PATH) -> SimulationDataType:
             Object containing simulation results.
         """
 
+        self.upload()
         self.start()
         self.monitor()
         return self.load(path=path)
 
-    @pd.root_validator()
-    def _upload(cls, values) -> None:
-        """Upload simulation to server without running."""
-
-        # task_id already present, don't re-upload
-        if values.get("task_id") is not None:
-            return values
+    @cached_property
+    def task_id(self):
+        upload_kwargs = {key: getattr(self, key) for key in self._upload_fields}
+        return web.upload(**upload_kwargs)
```

Comment:
why don't we use **self._upload_fields directly?

Reply:
`_upload_fields` is a tuple of strings referring to the keys. So we need to use `getattr()` to get the dictionary corresponding to the upload kwargs.

I agree this part is a bit ugly. I might rework the implementation a bit. The point, though, is that the `_upload_fields` class attribute defines what arguments must be passed to `web.upload()`.

 It was necessary to define this for adjoint, but now I'm trying to make use of it to generalize things a bit.

Reply:
did you try with instance_attrs = vars(obj)    or obj.__dict__  ?

Reply:
I think using pydantic `obj.dict().get(key)` will work. I'll try that.

Reply:
seemed to work fine with this approach